pkg/helpers/upgrader.go

```diff
@@ -0,0 +1,83 @@
+package helpers
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/pior/dad/pkg/config"
+	"github.com/pior/dad/pkg/executor"
+	"github.com/pior/dad/pkg/termui"
+)
+
+type Upgrader struct {
+	github  *Github
+	client  *http.Client
+	ui      *termui.UI
+	useSudo bool
+}
+
+// NewUpgrader returns a new upgrade helper using the default http client
+func NewUpgrader(cfg *config.Config, useSudo bool) (u *Upgrader) {
+	return NewUpgraderWithHTTPClient(cfg, http.DefaultClient, useSudo)
+}
+
+// NewUpgraderWithHTTPClient returns a new upgrade helper using the provided http client
+func NewUpgraderWithHTTPClient(cfg *config.Config, client *http.Client, useSudo bool) (u *Upgrader) {
+	g := NewGithubWithClient(cfg, client)
+	ui := termui.NewUI(cfg)
+
+	return &Upgrader{
+		github:  g,
+		client:  client,
+		useSudo: useSudo,
+		ui:      ui,
+	}
+}
+
+// Perform is fetching a new executable from `release`
+//   and upgrading the executable at `destinationPath` with it
+func (u *Upgrader) Perform(destinationPath string, sourceURL string) (err error) {
+	data, err := u.github.Get(sourceURL)
+
+	if err != nil {
+		return
+	}
+
+	tmpFile, err := makeTemporaryFile()
+	if err != nil {
+		return
+	}
+
+	if _, err = tmpFile.Write(data); err != nil {
+		return
+	}
+
+	cmdline := u.buildCmdline(tmpFile.Name(), destinationPath)
+
+	u.ui.CommandHeader(cmdline)
+
+	if _, err = executor.NewShell(cmdline).Run(); err != nil {
```

Comment:
If the command returns a non-zero code, `Run()` will return `<code>, nil`.
So here we would need to evaluate the code.

We may want to rethink the executor api after this PR...

Reply:
Is there any specific code we would need to evaluate in this specific case?